### PR TITLE
Update GET, POST, PUT APIs of subcategory and item

### LINKED ISSUE #68: Update the subcategory, item payloads to contain parent data
**Is your feature request related to a problem? Please describe.**
GET Subcategory payload should contain the category of that subcategory.
GET Item payload should contains the subcategories of that item.

**Describe the solution you'd like**
Remove the data that needs to be returned from @JsonIgnoreProperties annotation in respective classes

**Describe alternatives you've considered**
N/A

**Additional context**
Related thread: 
https://sef.discourse.group/t/design-forum-updating-academix-admin-panel/133/3

### LINKED ISSUE #69: Remove query parameters from item & subcategory post mappings
**Is your feature request related to a problem? Please describe.**
The post request of items has a query parameter to add subcategories.
The post request of the subcategory contains the category Id as a path param. 

**Describe the solution you'd like**
Add the parameters to body.

**Describe alternatives you've considered**
N/A

**Additional context**
Related thread:
https://sef.discourse.group/t/design-forum-updating-academix-admin-panel/133/3

### LINKED ISSUE #74: Can't update items, subcategories and category using PUT requests
**Describe the bug**
Only translations are updatable using item, subcategory, category put requests

**To Reproduce**
1. Check 'core/academix/items' PUT Request
2. See error(Only translations are updating)
*same error on subcategory and category

**Expected behavior**
Being fully updatable

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Ubuntu 19.04
 - Postman

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Piumal1999**
## Purpose

The purpose of this PR is to fix #69 
fix #68 
fix #74 

## Goals

Add category to subcategory GET payload
Add subcategories to item GET payload
Remove Query parameters from items POST request
Remove path parameters from subcategory POST request
Rewrite PUT requests

## Approach

Updated the models in order to change payloads
Updated the post requests to get categoryID and subcategoryIds from body
Rewrite subcategory, item put requests to update the parent

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

## Related PRs
 
N/A

## Test environment
 
Ubuntu 19.04
OpenJDK 1.8
Postman

## Learning

https://stackoverflow.com/questions/52656517/no-serializer-found-for-class-org-hibernate-proxy-pojo-bytebuddy-bytebuddyinterc

Related thread:
https://sef.discourse.group/t/design-forum-updating-academix-admin-panel/133/5

### CODE REVIEW

File: backend/src/main/java/org/sefglobal/core/academix/controller/admin/SubCategoryAdminController.java [hunk 1]
```diff
@@ -24,12 +24,11 @@ public SubCategoryAdminController(SubCategoryService subCategoryService) {
...
 
-    @PostMapping("/{categoryId}")
+    @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
-    public SubCategory addSubCategory(@PathVariable long categoryId,
-                                      @Valid @RequestBody SubCategory subCategory)
+    public SubCategory addSubCategory(@Valid @RequestBody SubCategory subCategory)
             throws ResourceNotFoundException {
-        return subCategoryService.addSubCategory(categoryId, subCategory);
+        return subCategoryService.addSubCategory(subCategory.getCategory().getId(), subCategory);
```
Comment (Gravewalker666) [inline backend/src/main/java/org/sefglobal/core/academix/controller/admin/SubCategoryAdminController.java:31]: Can't we just send the object itself to the addSubCategory method on service? Is it necessary to send the Id separately?

Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/controller/admin/SubCategoryAdminController.java]: Ah right. I'll update addSubCategory service

Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/controller/admin/SubCategoryAdminController.java]: Resolved

File: backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java [hunk 2]
```diff
@@ -97,31 +93,19 @@ public Category addCategory(Category category) {
...
-        if (!isUpdated) {
-            String msg = "Error, Category with id: " + id + " cannot be updated." +
-                         " Category doesn't exist.";
+    public Category updateCategory(long id, Category category) throws ResourceNotFoundException {
+        if (!categoryRepository.existsById(id)) {
+            String msg = "Error, Category with id: " + id + " cannot be updated. " +
+                    "Category doesn't exist.";
             log.error(msg);
             throw new ResourceNotFoundException(msg);
+        }else {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java:106]: Remove `else`

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java [hunk 3]
```diff
@@ -56,22 +52,25 @@ public Item getItemById(long id) throws ResourceNotFoundException {
...
      */
-    public Item addItem(List<Long> subCategoryIds, Item item) throws ResourceNotFoundException {
+    public Item addItem(Item item) throws ResourceNotFoundException {
+        List<Long> subCategoryIds = new ArrayList<>();
+        List<SubCategory> subCategories = item.getSubCategories();
+        for (SubCategory subCategory: subCategories){
+            subCategoryIds.add(subCategory.getId());
+        }
         List<SubCategory> existingSubCategories = subCategoryRepository.findAllById(subCategoryIds);
         if (subCategoryIds.size() > existingSubCategories.size()) {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:67]: You can remove the loop
```suggestion
        List<SubCategory> subCategories = item.getSubCategories();
        List<SubCategory> existingSubCategories = subCategoryRepository.findAllById(subCategoryIds);
        if (subCategories.size() > existingSubCategories.size()) {
```

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java [hunk 4]
```diff
@@ -81,31 +80,35 @@ public Item addItem(List<Long> subCategoryIds, Item item) throws ResourceNotFoun
...
-                })
-                .isPresent();
-        if (!isUpdated) {
-            String msg = "Error, Item with id: " + id + " cannot be updated. Item doesn't exist.";
+        if (!itemRepository.existsById(id)) {
+            String msg = "Error, Item with id: " + id + " cannot be updated. Item " +
+                    "doesn't exist.";
             log.error(msg);
             throw new ResourceNotFoundException(msg);
+        } else {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:96]: Remove `else`

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java [hunk 5]
```diff
@@ -81,31 +80,35 @@ public Item addItem(List<Long> subCategoryIds, Item item) throws ResourceNotFoun
...
+            List<Long> subCategoryIds = new ArrayList<>();
+            List<SubCategory> subCategories = item.getSubCategories();
+            for (SubCategory subCategory: subCategories){
+                subCategoryIds.add(subCategory.getId());
+            }
+            List<SubCategory> existingSubCategories = subCategoryRepository.findAllById(subCategoryIds);
+            if (subCategoryIds.size() > existingSubCategories.size()) {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:103]: You can remove the loop
```suggestion
            List<SubCategory> subCategories = item.getSubCategories();
            List<SubCategory> existingSubCategories = subCategoryRepository.findAllById(subCategoryIds);
            if (subCategories.size() > existingSubCategories.size()) {
```

File: backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java [hunk 6]
```diff
@@ -121,32 +118,30 @@ public SubCategory addSubCategory(long categoryId, SubCategory subCategory)
...
-                    return subCategoryRepository.save(updatableSubCategory);
-                })
-                .isPresent();
-        if (!isUpdated) {
+        if (!subCategoryRepository.existsById(id)) {
             String msg = "Error, SubCategory with id: " + id + " cannot be updated. SubCategory " +
                          "doesn't exist.";
             log.error(msg);
             throw new ResourceNotFoundException(msg);
+        }else {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java:133]: Remove `else`

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java (hunk 5 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:103]: But to get the list of subcategories, we need the list of subcategory Ids. 
`List<Long> subCategories`

Then the IDs taken from the request body are added to the List
Followed by the SubCategories relevant to each id are loaded into the `subCategories` List

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java (hunk 4 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:96]: Okay

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java (hunk 3 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:67]: But to get the list of subcategories, we need the list of subcategory Ids. 
`List<Long> subCategories`

Then the IDs taken from the request body are added to the List
Followed by the SubCategories relevant to each id are loaded into the `subCategories` List

File: backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java (hunk 2 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java:106]: Okay

File: backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java (hunk 6 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java]: Okay I'll update

File: backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java (hunk 2 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/CategoryService.java:106]: Resolved

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java (hunk 4 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java]: Resolved

File: backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java (hunk 6 above)
Comment (Piumal1999) [reply backend/src/main/java/org/sefglobal/core/academix/service/SubCategoryService.java]: Resolved

File: backend/src/main/java/org/sefglobal/core/academix/controller/admin/CategoryAdminController.java [hunk 7]
```diff
@@ -32,7 +32,7 @@ public Category addCategory(@Valid @RequestBody Category category) {
 
     @PutMapping("/{id}")
     @ResponseStatus(HttpStatus.ACCEPTED)
```
Comment (jayasanka-sack) [inline backend/src/main/java/org/sefglobal/core/academix/controller/admin/CategoryAdminController.java:34]: This should be `OK`

**Pull request merged**